src/H5Iint.c

```diff
@@ -9925,181 +9471,174 @@ H5I__discard_mt_type_info(H5I_mt_type_info_t * type_info_ptr)
 
-            assert( 0 < num_fl_entries_reallocable );
+        while ( ! done ) {
 
-            if ( atomic_compare_exchange_strong(&(H5I_mt_g.num_type_info_fl_entries_reallocable),
-                                                &num_fl_entries_reallocable, (num_fl_entries_reallocable - 1)) ) {
+            fl_shead = atomic_load(&(H5I_mt_g.type_info_fl_shead));
+            fl_stail = atomic_load(&(H5I_mt_g.type_info_fl_stail));
 
-                reallocable_entry_available = TRUE;
+            assert(fl_shead.ptr);
+            assert(fl_stail.ptr);
 
-            } else {
+            fl_snext = atomic_load(&(fl_shead.ptr->fl_snext));
 
-                /* update stats */
-                atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_num_reallocable_update_collisions), 1ULL);
-            }
-        } /* end while */
-#else
-        while ( ( ! reallocable_entry_available ) &&
-                ( 0 < (snum_fl_entries_reallocable = 
-                       atomic_load(&(H5I_mt_g.snum_type_info_fl_entries_reallocable))).val ) ) {
+            test_fl_shead = atomic_load(&(H5I_mt_g.type_info_fl_shead));
 
-            assert( 0 < snum_fl_entries_reallocable.val );
+            if ( ( test_fl_shead.ptr == fl_shead.ptr ) && ( test_fl_shead.sn == fl_shead.sn ) ) {
 
-            new_snum_fl_entries_reallocable.val = snum_fl_entries_reallocable.val - 1;
-            new_snum_fl_entries_reallocable.sn  = snum_fl_entries_reallocable.sn + 1;
+                if ( fl_shead.ptr == fl_stail.ptr ) {
 
-            if ( atomic_compare_exchange_strong(&(H5I_mt_g.snum_type_info_fl_entries_reallocable),
-                                                &snum_fl_entries_reallocable, new_snum_fl_entries_reallocable) ) {
+                    if ( NULL == fl_snext.ptr ) {
 
-                reallocable_entry_available = TRUE;
+                        /* the free list is empty */
+                        atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_frees_skipped_due_to_empty), 1);
+                        done = TRUE;
+                        break;
+                    }
 
-                fprintf(stderr, 
-                        "\ndisc: old/new snum_type_info_fl_entries_reallocable = {%lld, %lld} / {%lld, %lld}\n",
-                        (long long)snum_fl_entries_reallocable.val, 
-                        (long long)snum_fl_entries_reallocable.sn,
-                        (long long)new_snum_fl_entries_reallocable.val, 
-                        (long long)new_snum_fl_entries_reallocable.sn);
+                    /* attempt to set H5I_mt_g.type_info_fl_stail to fl_snext.  It doesn't
+                     * matter whether we succeed or fail, as if we fail, it
+                     * just means that some other thread beat us to it.
+                     *
+                     * that said, it doesn't hurt to collect stats
+                     */
+                    new_fl_stail.ptr = fl_snext.ptr;
+                    new_fl_stail.sn  = fl_stail.sn + 1;
```

Comment:
Same as for ID info, incrementing the SN by 1 might not agree with the actual SN of the next node.